util/src/main/java/com/redhat/lightblue/util/metrics/DropwizardRequestMetrics.java

```diff
@@ -0,0 +1,135 @@
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.util.metrics;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.Objects;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+public class DropwizardRequestMetrics implements RequestMetrics {
+    
+    private final Logger LOGGER = LoggerFactory.getLogger(DropwizardRequestMetrics.class);
+
+    private static final String API = "api";
+
+    private final MetricRegistry metricsRegistry;
+
+    public DropwizardRequestMetrics(MetricRegistry metricRegistry) {
+        metricsRegistry = metricRegistry;
+    }
+
+    /**
+     * Create exception namespace for metric reporting based on exception name
+     * 
+     */
+    private static String errorNamespace(String metricNamespace, Throwable exception) {
+        Class<? extends Throwable> actualExceptionClass = unravelReflectionExceptions(exception);
+        return name(metricNamespace, "requests", "exception", actualExceptionClass.getSimpleName());
+    }
+
+    /**
+     * Get to the cause we actually care about in case the bubbled up exception is a
+     * higher level framework exception that encapsulates the stuff we really care
+     * about.
+     * 
+     */
+    private static Class<? extends Throwable> unravelReflectionExceptions(Throwable e) {
+        while (e.getCause() != null
+                && (e instanceof UndeclaredThrowableException || e instanceof InvocationTargetException)) {
+            e = e.getCause();
+        }
+        return e.getClass();
+    }
+
+    @Override
+    public Context startEntityRequest(String operation, String entity, String version) {
+        return new DropwizardContext(name(API, operation, entity, version));
+    }
+
+    @Override
+    public Context startStreamingEntityRequest(String operation, String entity, String version) {
+        return new DropwizardContext(name(API, "stream", operation, entity, version));
+    }
+    
+    @Override
+    public Context startLockRequest(String lockOperation, String domain) {
+        return new DropwizardContext(name(API, "lock", domain, lockOperation));
+    }
+
+    @Override
+    public Context startHealthRequest() {
+        return new DropwizardContext(name(API, "checkHealth"));
+    }
+
+    @Override
+    public Context startDiagnosticsRequest() {
+        return new DropwizardContext(name(API, "checkDiagnostics"));
+    }
+
```

Comment:
Not sure if collecting metrics on /health and /diagnostics will be of much use. I don't think we would care about the timing of these requests or using these metrics to configure any alerts on these alert endpoints themselves.
There's no harm in keeping them (for consistency), but if they are not going to be used, it's better to keep these endpoints un monitored.

Reply:
The latencies won't be useful really since the health checks do the work in the background, the latencies aren't measuring any remote calls, only in-memory work which will take nanoseconds or less. But the counts are still useful I think, so we can see how often clients are asking about health or diagnostics, potentially spotting if load balancers are misconfigured or clients are asking way too often.